### PR TITLE
Ignoring empty statements in closures.

### PR DESCRIPTION
**Pull request by rscprof**
Solve #6116

### CODE REVIEW

File: tests/source/issue-6116/main.rs [hunk 1]
```diff
@@ -0,0 +1,7 @@
+fn foo() -> fn(i32) -> i32 {
+    |a| {
+        ;
+        a
+    }
+}
+
```
Comment (ytmimi) [inline tests/source/issue-6116/main.rs:7]: I'd like to expand on these test cases.

1. More than one statement in the block
    ```rust
    fn bar() -> fn(i32) -> i32 {
        |a| {
            ;
            a;
            b
        }
    }
    ```
2. Some tests where we have comments. For example:
   ```rust
   fn foo() -> fn(i32) -> i32 {
       |a| {
           /*comment before empty statement */;
           a
       }
   }
   ```
3. Any other edge cases that you can think of that we should consider.

**Review by ytmimi — CHANGES_REQUESTED** [review]
Thanks for another PR. I think we can simplify the implementation of `iter_stmts_without_empty` slightly, and I'd like to expand on the test cases that you've already added.

File: src/closures.rs [hunk 2]
```diff
@@ -1,3 +1,4 @@
+use rustc_ast::ast::StmtKind;
```
Comment (ytmimi) [inline src/closures.rs:1]: I didn't realize that `ast` was already brought into scope below. You might consider just using `ast::StmtKind` instead.

File: src/closures.rs [hunk 3]
```diff
@@ -116,14 +120,18 @@ fn get_inner_expr<'a>(
     expr
 }
 
+fn iter_stmts_without_empty(stmts: &[ast::Stmt]) -> impl Iterator<Item = &ast::Stmt> {
+    stmts.iter().filter(|x| !matches!(x.kind, StmtKind::Empty))
+}
```
Comment (ytmimi) [inline src/closures.rs:125]: nit: I'd really prefer if you used `stmt` instead of `x` in the closure, though if you want to use a single letter variable name then `s` feels more appropriate.

File: tests/source/issue-6116/main3.rs
Comment (ytmimi) [inline tests/source/issue-6116/main3.rs:1]: I think this test case is a great addition! Also wanted to note that you can totally continue to add new files as test cases, though you might find it easier to place all of these test cases into a single file.

**Review by ytmimi — COMMENTED** [review]
Thanks for adding those additional test cases, and for applying the other feedback. Please take a look at my follow up comments when you have a moment.

File: tests/source/issue-6116/main3.rs
Comment (rscprof) [reply tests/source/issue-6116/main3.rs:1]: I don't know what I have to do)))

Comment (ytmimi) [reply tests/source/issue-6116/main3.rs:1]: I'm letting you know that instead of creating multiple test files you can add more than one test case to a single file.

for example:

```rust
fn foo() -> fn(i32) -> i32 {
    |a| {
        ;
        a
    }
}

fn bar() -> fn(i32) -> i32 {
    |a| {
        ;
        a;
        b
    }
}

fn foobar() -> fn(i32) -> i32 {
    |a| {
        ;
	;
	;;;;
        a
    }
}

fn baz() -> fn(i32) -> i32 {
    |a| {
        /*comment before empty statement */;
        a
    }
}
```

**Comment by ytmimi** [comment]
@rscprof Thanks again for your help on this. I've taken another look at this PR and there's another test case that I'd like to see if we can resolve before merging.

Given the following input:

```rust
fn foo() -> fn(i32) -> i32 {
    |x| {
        /*comment before empty statement */;/*comment between statements */;
        x
    }
}
```

it formats as

```rust
fn foo() -> fn(i32) -> i32 {
    |x| {
        /*comment before empty statement */        /*comment between statements */
        x
    }
}
```

running rustfmt again formats as

```rust
fn foo() -> fn(i32) -> i32 {
    |x| {
        /*comment before empty statement */
 /*comment between statements */
        x
    }
}
```

and running rustfmt once again stabilizes as:

```rust
fn foo() -> fn(i32) -> i32 {
    |x| {
        /*comment before empty statement */
        /*comment between statements */
        x
    }
}
```
